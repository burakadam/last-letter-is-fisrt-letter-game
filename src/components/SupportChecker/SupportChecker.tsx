import { ISupportChecker } from '.';

declare global {
  interface Window {
    SpeechRecognition?: any;
    webkitSpeechRecognition?: any;
  }
}

const synth = window.speechSynthesis;

const speechRecognition =
  window.SpeechRecognition || window.webkitSpeechRecognition;

const SupportChecker = ({ children }: ISupportChecker) => {
  if (synth && speechRecognition) return <div>{children}</div>;
  return (
    <div className="w-screen h-screen flex justify-center items-center p-10">
      <p className="text-red-500 text-center">
        Tarayıcınız güncel ses ve mikrofon kütüphanelerini desteklemiyor. Lütfen
        daha güncel bir sürüm ile tekrar deneyiniz.
      </p>
    </div>
  );
};
export { SupportChecker };
