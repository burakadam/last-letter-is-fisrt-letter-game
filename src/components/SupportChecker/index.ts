export interface ISupportChecker {
  children: React.ReactNode;
}

export * from './SupportChecker';
