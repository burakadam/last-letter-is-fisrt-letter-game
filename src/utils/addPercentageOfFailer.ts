import { FAILER_PERCENTAGE } from '@/constants/failer';

// NOTE: shorstest way i found is to give 30% of failer chance is craete an array that has 30% of empty string

export const _addPercentageOfFailer = (list: string[]) => {
  const totalNumberOfEmptyArray = Math.floor(
    (list.length / (100 - FAILER_PERCENTAGE)) * FAILER_PERCENTAGE
  );

  const emptyArray = Array.from({ length: totalNumberOfEmptyArray }, () => '');

  const result = [...list, ...emptyArray];

  return result;
};
