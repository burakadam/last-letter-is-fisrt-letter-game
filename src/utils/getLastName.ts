export const _getLastName = (name: string) => name.split(' ').pop();
