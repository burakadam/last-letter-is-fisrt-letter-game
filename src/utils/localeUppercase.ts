import { DEFAULT_LANGUAGE } from '@/constants/language';

export const _localeUppercase = (name: string) =>
  name.toLocaleUpperCase(DEFAULT_LANGUAGE);
