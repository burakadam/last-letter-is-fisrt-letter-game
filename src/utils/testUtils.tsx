import { PreloadedState } from '@reduxjs/toolkit';
import { render, RenderOptions } from '@testing-library/react';
import React, { PropsWithChildren } from 'react';
import { Provider } from 'react-redux';

import { AppStore, RootState, setupStore } from '@/store/store';

interface ExtendedRenderOptions extends Omit<RenderOptions, 'queries'> {
  preloadedState?: PreloadedState<RootState>;
  store?: AppStore;
}

export function renderWithProviders(
  ui: React.ReactElement,
  {
    preloadedState = {},
    // Automatically create a store instance if no store was passed in
    store = setupStore(preloadedState),
    ...renderOptions
  }: ExtendedRenderOptions = {}
) {
  function Wrapper({ children }: PropsWithChildren<{}>): JSX.Element {
    return <Provider store={store}>{children}</Provider>;
  }
  return { store, ...render(ui, { wrapper: Wrapper, ...renderOptions }) };
}

export function Wrapper({ children }: { children: React.ReactElement }) {
  return <>{children}</>;
}

const customRender = (
  ui: React.ReactElement,
  options?: Omit<RenderOptions, 'wrapper'>
) => render(ui, { wrapper: Wrapper, ...options });

export * from '@testing-library/react';
export { customRender as render };
