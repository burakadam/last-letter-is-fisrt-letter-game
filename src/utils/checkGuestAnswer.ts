import NAMES from '@/data/names.json';
import { _localeUppercase } from './localeUppercase';

interface ICheckGuestAnswer {
  success: boolean;
  text: string;
}

export const _checkGuestAnswer = (
  guestAsnwer: string | null,
  hostAsnwer: string,
  answeredList: string[]
): ICheckGuestAnswer => {
  if (!guestAsnwer) return { success: false, text: 'Cavap veremedin' };

  const hostLastLetter = _localeUppercase(hostAsnwer).charAt(
    hostAsnwer.length - 1
  );

  const guestFirstLetter = _localeUppercase(guestAsnwer).charAt(0);

  if (guestFirstLetter !== hostLastLetter)
    return {
      success: false,
      text: `Verdiğin cevap "${hostLastLetter}" harfiyle başlamıyor`,
    };

  const isSaidBefore = answeredList.find(
    (name) => _localeUppercase(name) === _localeUppercase(guestAsnwer)
  );

  if (isSaidBefore)
    return {
      success: false,
      text: `${_localeUppercase(guestAsnwer)} ismini daha önce söylemiştin`,
    };

  const questAnswerIsAName = NAMES.filter(
    (name) => _localeUppercase(name) === _localeUppercase(guestAsnwer)
  );

  return {
    success: questAnswerIsAName.length > 0,
    text: 'Başarılı',
  };
};
