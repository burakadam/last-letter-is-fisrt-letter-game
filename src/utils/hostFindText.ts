import NAMES from '@/data/names.json';
import { _addPercentageOfFailer } from './addPercentageOfFailer';
import { _localeUppercase } from './localeUppercase';

// NOTE: look what said before and then check if valid after this function
export const _hostFindText = (
  saidList: string[],
  activeText: string | null
) => {
  let names = NAMES;

  if (activeText) {
    const lastLetter = _localeUppercase(activeText).charAt(
      activeText.length - 1
    );

    const filteredWithFirstLetter = NAMES.filter((name) =>
      _localeUppercase(name).startsWith(lastLetter)
    );

    if (!filteredWithFirstLetter) return '';

    const withoutSaidBefore = filteredWithFirstLetter.filter(
      (name) => !saidList.includes(name)
    );

    names =
      saidList.length > 3
        ? _addPercentageOfFailer(withoutSaidBefore)
        : filteredWithFirstLetter;
  }

  return names[Math.floor(Math.random() * names.length)];
};
