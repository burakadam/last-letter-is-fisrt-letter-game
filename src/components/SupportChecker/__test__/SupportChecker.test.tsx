import { render } from '@/utils/testUtils';

import { SupportChecker } from '../SupportChecker';

describe('render if browser supports SpeechRecognition or speechSynthesis', () => {
  it('render component correctly', () => {
    const { container } = render(<SupportChecker>App content</SupportChecker>);

    expect(container.firstChild).toMatchSnapshot();
  });
});
