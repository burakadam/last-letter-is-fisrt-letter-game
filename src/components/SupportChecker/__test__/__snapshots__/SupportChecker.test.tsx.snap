// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render if browser supports SpeechRecognition or speechSynthesis render component correctly 1`] = `
<div
  class="w-screen h-screen flex justify-center items-center p-10"
>
  <p
    class="text-red-500 text-center"
  >
    Tarayıcınız güncel ses ve mikrofon kütüphanelerini desteklemiyor. Lütfen daha güncel bir sürüm ile tekrar deneyiniz.
  </p>
</div>
`;
